### PR TITLE
8276126: Dump time class transformation causes heap objects of non-boot classes to be archived

### PR DESCRIPTION
**Pull request by calvinccheung**
During CDS static dump, if a class loaded by a boot class loader is transformed by a java agent, the transformed class is mistakenly identified as a class loaded by a custom loader. This would cause archiving of the heap objects to fail.

The proposed patch corrects the identification of the transformed class and will not include it in the archive.
Also, if a transformed class is detected during dump time, the archiving of the heap objects will be disabled.

Testing: Oracle CI tiers 1-3, tier4 in progress.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276126](https://bugs.openjdk.java.net/browse/JDK-8276126): Dump time class transformation causes heap objects of non-boot classes to be archived

### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to 381f26d6950b45a4cbd03d1cfeba9897f332be6f
 * [Yumin Qi](https://openjdk.java.net/census#minqi) (@yminqi - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6484/head:pull/6484` \
`$ git checkout pull/6484`

Update a local copy of the PR: \
`$ git checkout pull/6484` \
`$ git pull https://git.openjdk.java.net/jdk pull/6484/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6484`

View PR using the GUI difftool: \
`$ git pr show -t 6484`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6484.diff">https://git.openjdk.java.net/jdk/pull/6484.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/classfile/classLoaderExt.cpp [hunk 1]
```diff
@@ -227,7 +229,8 @@ void ClassLoaderExt::setup_search_paths(JavaThread* current) {
   ClassLoaderExt::setup_app_search_path(current);
 }
 
-void ClassLoaderExt::record_result(const s2 classpath_index, InstanceKlass* result) {
+void ClassLoaderExt::record_result(const s2 classpath_index, InstanceKlass* result,
+                                   const ClassFileStream* stream, bool redefined) {
```
Comment (iklam) [inline src/hotspot/share/classfile/classLoaderExt.cpp:233]: The `stream` argument is not used.

**Review by iklam — APPROVED** [review]
LGTM. One small nit about unused parameter.

File: src/hotspot/share/classfile/classLoaderExt.cpp (hunk 1 above)
Comment (calvinccheung) [reply src/hotspot/share/classfile/classLoaderExt.cpp]: Thanks for spotting this. I've fixed it.

File: src/hotspot/share/classfile/classLoaderExt.cpp [hunk 2]
```diff
@@ -245,4 +246,21 @@ void ClassLoaderExt::record_result(const s2 classpath_index, InstanceKlass* resu
   }
   result->set_shared_classpath_index(classpath_index);
   result->set_shared_class_loader_type(classloader_type);
+#if INCLUDE_CDS_JAVA_HEAP
+  if (DumpSharedSpaces && AllowArchivingWithJavaAgent && classloader_type == ClassLoader::BOOT_LOADER &&
```
Comment (yminqi) [inline src/hotspot/share/classfile/classLoaderExt.cpp:250]: the function is call during dump time (the assert at first line of this function check that) so DumpSharedSpace can be removed.
I am a little confused by this part:  If 'redefined' is 'true', The 'redefined' only should be enough for set disable_writing() whether other conditions stands or not?

Comment (calvinccheung) [reply src/hotspot/share/classfile/classLoaderExt.cpp:250]: The `Arguments::assert_is_dumping_archive()` includes both static and dynamic dumping. The archiving of heap objects is performed only during static dumping. So the check for `DumpShareSpaces` is needed.

The `redefined` flag alone is not enough to disable writing heap objects to archive. If a class loaded by an AppClassLoader and is redefined, it should not disable archiving heap objects. We currently have some test cases under runtime/cds/appcds/javaldr to test that.

Comment (yminqi) [reply src/hotspot/share/classfile/classLoaderExt.cpp:250]: > 

Thanks for the explanation.

**Review by yminqi — APPROVED** [review]
LGTM.

**Comment by calvinccheung** [comment]
Thanks @iklam and @yminqi for the review.
/integrate

**Pull request closed without merging**